src/upload.ts

```diff
@@ -0,0 +1,150 @@
+  let uploadHandler = config.uploadHandler || defaultUpload;
+  let sequenceNo = raw.envelope.sequenceNumber;
+  let onSuccess = (status: number) => {
+    onSuccessDefault(sequenceNo, status);
+    if (onSuccessExtra) {
+      onSuccessExtra(status);
+    }
+  };
+  let onFailure = (status: number) => {
+    onFailureDefault(sequenceNo, status);
+    if (onFailureExtra) {
+      onFailureExtra(status);
+    }
+  };
+  uploadInfos[sequenceNo] = { compressed, raw, failureCount: 0 };
+  uploadHandler(compressed, onSuccess, onFailure);
+
+  if (config.debug) {
+    let rawLength = JSON.stringify(raw).length;
+    debug(`** Upload #${uploadCount}: Batch #${sequenceNo}`
+          + ` ${Math.round(compressed.length / 1024.0)}KB`
+          + ` (${Math.round(rawLength / 1024.0)}KB Raw). **`);
+  }
+  uploadCount++;
+}
+
+export function enqueueUpload(compressed: string, raw: IPayload) {
+  let payloadInfo: IUploadInfo = { compressed, raw, failureCount: 0 };
+  uploadQueue.push(payloadInfo);
+}
+
+export function flushUploadQueue() {
+  uploadSequentially(uploadQueue);
+}
+
+export function uploadSequentially(uploads: IUploadInfo[]) {
+  if (uploads.length > 0) {
+    let nextUpload = uploads.shift();
+    let uploadNext = (status: number) => {
+      uploadSequentially(uploads);
+    };
+    upload(nextUpload.compressed, nextUpload.raw, uploadNext, uploadNext);
+  }
+}
+
+export function resetUploads() {
+  sentBytesCount = 0;
+  uploadCount = 0;
+  uploadQueue = [];
+  uploadInfos = {};
+  reUploadQueue = [];
+}
+
+function defaultUpload(payload: string, onSuccess?: UploadCallback, onFailure?: UploadCallback) {
+  if (config.uploadUrl.length > 0) {
+    payload = JSON.stringify(payload);
+    let xhr = new XMLHttpRequest();
+    xhr.open("POST", config.uploadUrl);
+    xhr.setRequestHeader("Content-Type", "application/json");
+    xhr.onreadystatechange = () => { onXhrReadyStatusChange(xhr, onSuccess, onFailure); };
+    xhr.send(payload);
+  }
+}
+
+function onXhrReadyStatusChange(xhr: XMLHttpRequest, onSuccess: UploadCallback, onFailure: UploadCallback) {
+  if (xhr.readyState === XMLHttpRequest.DONE) {
+    // HTTP response status documentation:
+    // https://developer.mozilla.org/en-US/docs/Web/HTTP/Status
+    if (xhr.status < 200 || xhr.status > 208) {
+      onFailure(xhr.status);
+    } else {
+      onSuccess(xhr.status);
+    }
+  }
+}
+
+function onSuccessDefault(sequenceNumber: number, status: number) {
+  debug(`SUCCESS: Delivered batch #${sequenceNumber}`);
+  let sentData = uploadInfos[sequenceNumber];
+  delete uploadInfos[sequenceNumber];
+  sentBytesCount += sentData.compressed.length;
+
+  if (state === State.Activated && sentBytesCount > config.totalLimit) {
+    let totalByteLimitExceededEventState: ITotalByteLimitExceededEventState = {
+      type: Instrumentation.TotalByteLimitExceeded,
+      bytes: sentBytesCount
+    };
+    instrument(totalByteLimitExceededEventState);
+    teardown();
+    return;
+  }
+
+  reUpload();
```

Comment:
>reUpload(); [](start = 2, length = 11)

why don't you have the reUpload() call in onFailure instead? Presumably in most cases this is going to be a noop from the success path and the failure path is really when you'd be interested in retrying right?

Reply:
Presumably, if an xhr failed, there is something wrong with the connection. When the next xhr is successful, we know that network is OK and we can now attempt reuploading the failed batch.

---
In reply to: [168000770](https://github.com/Microsoft/clarity-js/pull/63#discussion_r168000770) [](ancestors = 168000770)